File: python/Arrays&Hashing/longest_consecutive_sequence.py
from typing import List

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        """
        Approach:
        - Use a set to allow O(1) lookups for each number.
        - For each number, check if it is the start of a sequence (num - 1 not in set).
        - If so, count the length of the consecutive sequence starting from that number.
        - Track the maximum sequence length found.

        Time Complexity: O(n), where n is the number of elements in nums (each number is processed at most twice).
        Space Complexity: O(n), for the set storing all numbers.
        """
        numSet = set(nums)
        max_size = 0  # Initialize to 0 for empty array case
        
        for num in numSet:
            # Check if this is the start of a sequence
            if num - 1 not in numSet:
                curr_size = 1
                # Count how long the sequence is
                while num + 1 in numSet:
                    curr_size += 1
                    num += 1
                max_size = max(max_size, curr_size)
                
        return max_size

# -------------------
# Test Cases
# -------------------
def test_longest_consecutive_sequence():
    sol = Solution()
    # Test case 1: Example from Leetcode
    nums1 = [100, 4, 200, 1, 3, 2]
    assert sol.longestConsecutive(nums1) == 4  # Sequence: 1,2,3,4
    # Test case 2: No consecutive sequence
    nums2 = [10, 5, 100]
    assert sol.longestConsecutive(nums2) == 1
    # Test case 3: Empty list
    nums3 = []
    assert sol.longestConsecutive(nums3) == 0
    # Test case 4: All elements consecutive
    nums4 = [1, 2, 3, 4, 5]
    assert sol.longestConsecutive(nums4) == 5
    # Test case 5: Duplicates in input
    nums5 = [1, 2, 2, 3]
    assert sol.longestConsecutive(nums5) == 3
    print("All test cases passed!")

if __name__ == "__main__":
    test_longest_consecutive_sequence()

File: C++/Arrays&Hashing/longest_consecutive_sequence.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Find longest consecutive sequence in array
     * 
     * Approach:
     * - Store numbers in hash set for O(1) lookup
     * - For each number, check if it's start of sequence
     * - If it is (num-1 not in set), count sequence length
     * - Track maximum sequence length found
     * 
     * Time Complexity: O(n) - each number visited at most twice
     * Space Complexity: O(n) - storing all numbers in set
     */
    int longestConsecutive(vector<int>& nums) {
        // Create hash set from input array
        unordered_set<int> numSet(nums.begin(), nums.end());
        int maxSize = 0;
        
        // Check each possible sequence
        for (int num : numSet) {
            // Only process numbers that could start a sequence
            if (numSet.find(num - 1) == numSet.end()) {
                int currSize = 1;
                int currNum = num;
                
                // Count consecutive numbers
                while (numSet.find(currNum + 1) != numSet.end()) {
                    currSize++;
                    currNum++;
                }
                
                maxSize = max(maxSize, currSize);
            }
        }
        
        return maxSize;
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Header and Namespace:
 *    #include <bits/stdc++.h>    // Include all standard libraries
 *    using namespace std;        // Makes std:: prefix optional
 * 
 * 2. Function Declaration:
 *    int longestConsecutive(     // Return type is integer
 *        vector<int>& nums)      // Takes reference to vector of integers
 * 
 * 3. Set Creation:
 *    unordered_set<int> numSet(nums.begin(), nums.end())
 *    - unordered_set: hash set container
 *    - nums.begin(), nums.end(): iterator range for initialization
 *    - O(1) lookup time
 * 
 * 4. Range-based For Loop:
 *    for (int num : numSet)
 *    - Iterates through each number in set
 *    - int num: creates copy of current element
 * 
 * 5. Set Operations:
 *    numSet.find(num - 1)       // Search for element
 *    numSet.end()               // Iterator to end of set
 *    == comparison checks if element not found
 * 
 * 6. Variables and Scope:
 *    int currSize = 1           // Local variable initialization
 *    int currNum = num          // Copy value to new variable
 * 
 * 7. While Loop:
 *    while (condition)          // Repeat while condition is true
 *    currNum++                  // Increment operator
 * 
 * 8. Standard Library Function:
 *    max(maxSize, currSize)     // Returns larger of two values
 * 
 * Note: This implementation achieves:
 * - O(n) time complexity by visiting each number at most twice
 * - O(n) space complexity for storing the hash set
 */
